### ISSUE TITLE
IFindSagas implementation example for Azure Storage persistance

### ISSUE THREAD

**Issue by DixonDs** [Particular/NServiceBus.Persistence.AzureStorage#180]
According to docs at https://docs.particular.net/nservicebus/sagas/saga-finding:
>// SynchronizedStorageSession will have a persistence specific extension method

However, I have troubles finding any extension method relevant to Table Storage persistence. Is there any example how to implement IFindSagas for Azure Storage persistence?

**Comment by yvesgoeleven** [comment]
@DixonDs Querying azure storage over anything but the partition rowkey pair will lead to super bad performance in the long run. Therefore it is not supported to use saga finding based on other properties. If you need this feature I suggest to consider another data store for saga's e.g. Sql Database

**Comment by DixonDs** [comment]
@yvesgoeleven In my particular use-case, I need to implement a singleton saga without adding a dummy correlation property to all message types that can start this saga. So, I would still query by partition key for that matter.

**Comment by yvesgoeleven** [comment]
@DixonDs saga id is the partition key, so the default lookup should already work for your use case then.

**Comment by DixonDs** [comment]
@yvesgoeleven Yes, I know, so back to the original question. Is there any extension method to `SynchronizedStorageSession` specific to Azure Storage persistence that would allow me to get the saga data by saga ID in my custom `IFindSagas` implementation?

**Comment by yvesgoeleven** [comment]
@DixonDs no, there is no such extension method

**Comment by boblangley** [comment]
@DixonDs Due to the nature of the persistence, and the limited indexing capabilities of Azure Storage Tables, we do not think implementing a `SynchronizedStorageSession` is feasible for most customers. I would happy to talk to you 1 on 1 about your specific use-case. Would you mind opening a ticket with `support@particular.net` so we can chat specifics?

**Comment by dasjestyr** [comment]
I'm a bit confused, are you saying that one cannot do custom saga finders at all if we're using azure storage persistence? I have a similar case, it seems. I have to a few steps to determine the ID from the message (hence the need for the custom finder), but ultimately, it's just the same saga ID. I don't have multiple IDs or anything, so I'm not sure what the problem is.

**Comment by yvesgoeleven** [comment]
@dasjestyr what you are looking for is the mapping between message properties and saga id, which can be done using https://docs.particular.net/nservicebus/sagas/?version=core_7#a-simple-saga, you don't need the saga finder for it. That should work find with azure storage.

It is correct that you cannot implement custom `IFindSagas` because of the limits that azure storage has on querying by field. Eventhough the names are fairly similar, both features serve a different purpose.

**Comment by dasjestyr** [comment]
I ended up removing the saga for now (it was very short), but the problem is that the event that comes back from the other service contains the ID I need in a metadata field instead of a property. So I need to actually dig it out of a dictionary and then convert it to a GUID  before mapping it, but the mapper interface for sagas are simply expression -> expression. I was able to put all of the casting and conversions into a single statement to fit the expression (and it worked), but that's just ... gross. This is why I figured maybe the custom finder would have been better. Would there have been a better way?

**Comment by Antwane** [comment]
I'm having the same problem, I need to correlate based on one of the header of the message not one of the properties and I'm using table storage. So its not possible to correlate by header because the context of the message is not included in ConifgurationMapper and is not possible to create your own saga finder because Table storage is not supported that leaves me in dead water.

**Comment by yvesgoeleven** [comment]
@dasjestyr you can extract that expression into a method that returns the expression, that will make it better.

@Antwane I'll raise an issue for you that the ConfigureHowToFindSaga logic doesn't provide access to the headers, that's what you need.

We cannot add a GetSession() like extension to SynchronizedStorageSession, as that would imply there is actually a context that can be shared. Azure storage does not support that, every operation is non transactional and is standalone, there is nothing to share. 

Next to that it's too dangerous to allow querying on anything else but the partitionkey. But thinking of it, there is nothing preventing you from creating a cloud storage account instance in an implementation of `IFindSagas`, read the data from the table and return it. I highly recommend against it though.

**Comment by Antwane** [comment]
@yvesgoeleven Thank that feature would be much appreciated.

After a lot time spent on this the only option is to query Table Storage directly, I understand you are advising against it because I would be relying on particular implementation of Azure table storage and that might change with greater version, at the moment is the only option I see.
